### PR TITLE
Handle dynamic init IO

### LINKED ISSUE #34: Current implementation would straight up crash on Arduino API calls at dynamic init
(Issue opened here since the discussion will have a direct impact on whether we implement a fix or not)
@platisd I can't help but notice that in your examples for `smartcar_shield` at [this location](https://github.com/platisd/smartcar_shield/blob/master/examples/sensors/odometer/singleOdometer/singleOdometer.ino#L6) you declare a global variable which gets initialized at dynamic init (https://eel.is/c++draft/basic.start#dynamic), which results in [this other location](https://github.com/platisd/smartcar_shield/blob/master/src/sensors/odometer/interrupt/DirectionlessOdometer.cpp#L27) to be run, itself causing [this last location](https://github.com/platisd/smartcar_shield/blob/master/src/runtime/arduino_runtime/ArduinoRuntime.cpp#L17) to invoke the Arduino IO API function `pinMode`.

Now, does the Arduino framework guarantee that IO API functions will work as expected when invoked at dynamic init?
If not, is it simply always the case despite no guarantees from the framework?
The C++ standard only mandates expected behavior for its iostreams (which aren't required on standalone implementations like often found on embedded anyway).

The main question boils down to: while implementation-defined regarding the C++ standard, is the support for hardware IO API calls at dynamic initialization widespread and relied-upon enough that it would be reasonable and more or less expected of us to go out of our way to support that scenario?

Thanks in advance for your time and your input; your experience with embedded is precious

[While we're at it, since I dug a bit on how you manage your runtime global in `smartcar_shield`, I ought to inform you that if someone were to write for a single program multiple C++ translation units which all include `<Smartcar.h>` to get access to your library headers, they would get multiple definitions of `arduinoRuntime` therefore breaking ODR and therefore failing to link; you didn't document this anywhere, so I assume it's a bug.]

### PR DESCRIPTION
**Pull request by AeroStun**
Working partial fix for #34
Closes #34

Allows for calls to IO functions at dynamic init
Caveats:
- Any IO on pins that do not exist is discarded before call to `setup`
- Any IO on pins above 255 is immediately discarded
- Error policy is always silent during dynamic init

### CODE REVIEW

File: modules/ardrivo/src/Entrypoint.cxx [hunk 1]
```diff
@@ -16,15 +16,70 @@
...
+    loc_board_data->uart_buses = std::vector<UartBus>(64);
+    loc_board_data->interrupts_handlers = decltype(board_data->interrupts_handlers)(255);
+    board_data = loc_board_data.release();
+
+    auto loc_board_info = std::make_unique<BoardInfo>();
+    loc_board_info->pins_caps.resize(255);
+    ranges::for_each(loc_board_info->pins_caps, [](PinCapability& pin){ pin.digital_in = pin.digital_out = true; }); // default to all digital-able
+    // Protocol-pins need not be set as we cannot do pin-locking detection during dynamic-init anyway
+
+    ::board_info = loc_board_info.release();
```
Comment (RuthgerD) [inline modules/ardrivo/src/Entrypoint.cxx:65]: Cant this just be just `board_info` without the `::`?

File: modules/ardrivo/src/Entrypoint.cxx [hunk 2]
```diff
@@ -16,15 +16,70 @@
...
     board_info = info;
     return true;
+}
+
+void maybe_init() noexcept {
+    if (board_data)
+        return;
+
+    // over-allocate everything to handle possible IO at dynamic-init
+    auto loc_board_data = std::make_unique<BoardData>();
```
Comment (RuthgerD) [inline modules/ardrivo/src/Entrypoint.cxx:49]: `std::make_unique` vs direct `std::unique_ptr` ?

Comment (RuthgerD) [reply modules/ardrivo/src/Entrypoint.cxx:49]: Im just stupid

File: modules/ardrivo/src/Entrypoint.cxx (hunk 1 above)
Comment (AeroStun) [reply modules/ardrivo/src/Entrypoint.cxx:65]: It can, but I want to make it clear that it's global scope

**Pull request merged**